examples/cpp/keyvaluestore/caching_interceptor.h

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <map>
+
+#include <grpcpp/support/client_interceptor.h>
+
+#ifdef BAZEL_BUILD
+#include "examples/protos/keyvaluestore.grpc.pb.h"
+#else
+#include "keyvaluestore.grpc.pb.h"
+#endif
+
+// This is a naive implementation of a cache. A new cache is for each call. For
+// each new key request, the key is first searched in the map and if found. Only
+// if the key is not found in the cache do we make a request.
+class CachingInterceptor : public grpc::experimental::Interceptor {
+ public:
+  CachingInterceptor(grpc::experimental::ClientRpcInfo* info) {}
+
+  void Intercept(
+      ::grpc::experimental::InterceptorBatchMethods* methods) override {
+    bool hijack = false;
+    if (methods->QueryInterceptionHookPoint(
+            grpc::experimental::InterceptionHookPoints::
+                PRE_SEND_INITIAL_METADATA)) {
+      // Hijack all calls
+      hijack = true;
```

Comment:
As far as the interface goes, I don't understand why we have a limitation that we can only hijack at pre_send_initial_metadata. I guess the idea is that then the server has not gotten any day, but at any given point after we hijack, we could send partial data over to the server anyway, so that's not the argument.

Reply:
At pre_send_initial_metadata, the interceptor should have all information to decide whether it needs to hijack or not. Having it hijack later means that we need to take care of doing some cleanup since the server has probably already done some work.. and that seems unnecessary